src/ngx_http_lua_kong_module.c

```diff
@@ -407,42 +431,78 @@ ngx_http_lua_kong_cleanup(void *data)
         sk_X509_pop_free(ctx->upstream_client_certificate_chain, X509_free);
         EVP_PKEY_free(ctx->upstream_client_private_key);
     }
+
+    if (ctx->upstream_trusted_store != NULL) {
+        X509_STORE_free(ctx->upstream_trusted_store);
+    }
 }
 
+static void
+ngx_http_lua_kong_cleanup_cert_and_key(void *data)
+{
+    ngx_http_lua_kong_ctx_t     *ctx = data;
+
+    if (ctx->upstream_client_certificate_chain != NULL) {
+        sk_X509_pop_free(ctx->upstream_client_certificate_chain, X509_free);
+        EVP_PKEY_free(ctx->upstream_client_private_key);
+    }
+}
 
-int
-ngx_http_lua_kong_ffi_set_upstream_client_cert_and_key(ngx_http_request_t *r,
-    void *_chain, void *_key)
+static void
+ngx_http_lua_kong_cleanup_trusted_store(void *data)
 {
-    STACK_OF(X509)              *chain = _chain;
-    EVP_PKEY                    *key = _key;
-    STACK_OF(X509)              *new_chain;
-    ngx_http_lua_kong_ctx_t     *ctx;
-    ngx_pool_cleanup_t          *cln;
+    ngx_http_lua_kong_ctx_t     *ctx = data;
 
-    if (chain == NULL || key == NULL) {
-        return NGX_ERROR;
+    if (ctx->upstream_trusted_store != NULL) {
+        X509_STORE_free(ctx->upstream_trusted_store);
     }
+}
+
+ngx_http_lua_kong_ctx_t *
+ngx_http_lua_kong_get_module_ctx(ngx_http_request_t *r)
+{
+    ngx_http_lua_kong_ctx_t     *ctx;
+    ngx_pool_cleanup_t          *cln;
 
     ctx = ngx_http_get_module_ctx(r, ngx_http_lua_kong_module);
     if (ctx == NULL) {
         ctx = ngx_pcalloc(r->pool, sizeof(ngx_http_lua_kong_ctx_t));
         if (ctx == NULL) {
-            return NGX_ERROR;
+            return NULL;
         }
 
         cln = ngx_pool_cleanup_add(r->pool, 0);
         if (cln == NULL) {
-            return NGX_ERROR;
+            return NULL;
         }
 
         cln->data = ctx;
         cln->handler = ngx_http_lua_kong_cleanup;
 
         ngx_http_set_ctx(r, ctx, ngx_http_lua_kong_module);
+    }
+    return ctx;
+}
+
 
+int
+ngx_http_lua_kong_ffi_set_upstream_client_cert_and_key(ngx_http_request_t *r,
+    void *_chain, void *_key)
+{
+    STACK_OF(X509)              *chain = _chain;
+    EVP_PKEY                    *key = _key;
+    STACK_OF(X509)              *new_chain;
+    ngx_http_lua_kong_ctx_t     *ctx;
+
+    if (chain == NULL || key == NULL) {
+        return NGX_ERROR;
+    }
+
+    ctx = ngx_http_lua_kong_get_module_ctx(r);
+    if (ctx == NULL) {
+        return NGX_ERROR;
```

Comment:
Style: add a newline after this line. Nginx C source code always add a newline before `else if` or `else`.